PyDynamic/model_estimation/fit_filter.py

```diff
@@ -87,92 +91,224 @@ def _fitIIR(
+    if verbose:
+        monte_carlo_message = (
+            f" Uncertainties of the filter coefficients are "
+            f"evaluated using the GUM S2 Monte Carlo method "
+            f"with {mc_runs} runs."
+        )
+        print(
+            f"LSIIR: Least-squares fit of an order {max(Nb, Na)} digital IIR filter to"
+            f"{' the reciprocal of' if inv else ''} a frequency response "
+            f"given by {len(Hvals)} values.{monte_carlo_message if UHvals else ''}"
+        )
 
-    while stable is not True and run < maxiter:
-        g1 = grpdelay(b, a, Fs)[0]
-        g2 = grpdelay(b, astab, Fs)[0]
-        tau = np.ceil(tau + np.median(g2 - g1))
+    # Initialize the warning message in case the final filter will still be unstable.
+    warning_unstable = "CAUTION - The algorithm did NOT result in a stable IIR filter!"
 
-        b, a = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=False)
-        if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-            astab = mapinside(a)
+    # Prepare frequencies, fitting and stabilization parameters.
+    w = 2 * np.pi * f / Fs
+    Ns = np.arange(0, max(Nb, Na) + 1)[:, np.newaxis]
+    E = np.exp(-1j * np.dot(w[:, np.newaxis], Ns.T))
+    as_and_bs = np.empty((mc_runs, Nb + Na + 1))
+    taus = np.zeros((mc_runs,), dtype=int)
+    stab_iters = np.zeros((mc_runs,), dtype=int)
+
+    # Conduct the Monte Carlo runs or in case we did not have uncertainties execute
+    # just once the actual algorithm.
+    for mc_run in range(mc_runs):
+        # Conduct actual fit.
+        b_i, a_i = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=inv)
+
+        # Determine if the computed filter already is stable.
+        unstable = not check_stability(a_i)
+
+        # In case the user specified not to check for stability and stabilize the filter
+        # if necessary, we just return the result so far and inform about hte result.
+        if justFit:
+            if verbose:
+                sos = np.sum(np.abs((dsp.freqz(b_i, a_i, w)[1] - Hvals) ** 2))
+                print(
+                    f"LSIIR: Fitting "
+                    f"{'' if UHvals is None else f'for MC run {mc_run} '}"
+                    f"finished. No stabilization requested"
+                    f". {warning_unstable if unstable else ''} "
+                    f"Final sum of squares = {sos}"
+                )
+            # Since no stabilization was done, we return tau = 0 regardless of the
+            # desired initial estimate of the time delay provided for the stabilization.
+            taus[mc_run] = 0
+            continue
         else:
-            stable = True
-        run = run + 1
+            # Otherwise we initialize the time delay for the current Monte Carlo run.
+            taus[mc_run] = tau
+
+        # Initialize counter which we use to report about required iteration count.
+        current_stab_iter = 0
+        # Stabilize filter coefficients with a maximum of 50 iterations.
```

Comment:
This comment was written, when during the rewrite  max_stab_iter did not exist and the limit was just hardcoded. We should adapt it.

```suggestion
        # Stabilize filter coefficients with a maximum number of iterations as specified.
```